tools/src/main/scala/scala/scalanative/linker/Assembly.scala

```diff
@@ -16,14 +14,16 @@ object Assembly {
   final case class Dir private[Assembly](val base: File) extends Assembly {
     //println(s"discovered dir assembly $base")
 
+    def matchFiles(f: File): Array[File] = {
+      if (f.isDirectory) f.listFiles.map(matchFiles).flatten.toArray
+      else if (f.getName.endsWith(".nir")) Array(f)
+      else Array()
+    }
```

Comment:
A naive use of lists would also be quadratic. But the following is linear:
```scala
    def matchFiles(f: File, tail: List[File] = Nil): Array[File] = {
      if (f.isDirectory) f.listFiles.foldRight(tail)(matchFiles)
      else if (f.getName.endsWith(".nir")) f :: tail
      else tail
    }
```
"Typical" difference list. :p